### ISSUE TITLE
Add Profiling-API callback for AssemblyBindFailure's

### ISSUE THREAD

**Issue by discostu105** [dotnet/runtime#5429]
Here is my use-case:
- I'm trying to write a Profiler for the CoreCLR (which is hooked into the process via Profiling API (CORECLR_ENABLE_PROFILING, ...)
- It's intercepting ModuleLoad and JitCompilationStarted callbacks to modify Module Metadata on the fly and inject IL code into certain methods.
- The injected IL instructions call a managed helper method, which is defined in a separate assembly (`MyProfiler.dll`, `MyProfiler.MethodEnter()`)
- In order to inject that IL instruction, I need to add a TypeRef/MemberRef for `MyProfiler.MethodEnter()` and an AssemblyRef to `MyProfiler.dll` as resolution scope. The AssemblyRef would point to `MyProfiler` with a certain version, hashcode, ect.
- Since the AssemblyRef to `MyProfiler.dll` is injected at runtime, that assembly is _not_ referenced by the application at build-time. So, there is nothing about it in `[AssemblyName].deps`

My question is: How can I make the CoreCLR resolve that AssemblyRef? I already figured, Assemblies are searched in the application directory and in the "Trusted Platform Assemblies". If I understood that right, these are passed in by the [corehost](https://github.com/borgdylan/dotnet-cli/blob/master/Documentation/specs/corehost.md), and usually they come from the global nuget package cache.

So, I see three possible solutions for this:
 (1) add `MyProfiler.dll` to the nuget package cache on the machine
 (2) somehow add `MyProfiler.dll` to the TPA list (is that possible from within a profiler?)
 (3) Add a Profiling-API callback like `AssemblyBindFailure` and allow the profiler to specify a path to the dll. The CLR would then try to bind that assembly once more with the given path.

I'm not very much in favor of (1), because it would involve some "installer-step" for my profiler to get it to work. I thought (3) is a neat idea that would solve my problem elegantly. What do you think about it?

Or is there a much simpler solution to my problem?

**Comment by Mahendra95-sys** [comment]
I am trying to Instrument .NET Core web applications that runs on .NET Core 3.1 using CoreCLR Profiler.
I have set the environment values CORECLR_PROFILER , CORECLR_ENABLE_PROFILING  and CORECLR_PROFILER_PATH, where my CoreCLRProfiler dll gets attached to dotnet.exe and it is getting the callbacks.

I am able to get all the callbacks,but when i allow injecting the code into the Webapplication's method then the app is getting crashed(dotnet.exe gets killed) as it couldnt find the injected function call.

I have created helper assembly(.NET standard 2.0) with the injected functions body and signed it with strong name and installed it in to the GAC. And also used DefineAssemblyRef(),DefineTypeRefByName() and DefineMemberRef() from IMetaDataAssemblyEmit to load assembly and its class methods. And also tried by placing dotnet standard dll in application folder. But the helper assembly is not loaded to dotnet.exe process.

Where should my helper assembly placed..? and

how can I load helper assembly to dotnet process from my native coreclr profiler?

It would be much helpful if i get some correct direction to load or use helper assembly to dotnet process.

I have tried using  httpss://github.com/caozhiyuan/ClrProfiler.Trace but that profiler Is not loaded into dotnet process.
What "hook il in premian" mean? 

How httpss://github.com/caozhiyuan/ClrProfiler.Trace example is loading managed assembly in to dotnet.exe from Native code?
